### ISSUE TITLE
can not catch Assert in gdb

### ISSUE THREAD

**Issue by tjhei** [dealii/dealii#4490]
With the change from abort() to MPI_Abort() in #4413 I can no longer break at the point of the assert in gdb. Any suggestions on how to fix/work around this? 

One could detect if we are running inside a debugger (but this is somewhat ugly) and do abort() instead. Or maybe listen to some ENV variable?

**Comment by bangerth** [comment]
Oh, what the hell :-( What happens if you run into `MPI_Abort()` when in a debugger?

**Comment by tjhei** [comment]
It just ends the process unconditionally instead of breaking in the abort.

**Comment by bangerth** [comment]
Hm. What do we do?

**Comment by Rombur** [comment]
Can't we add something like `b MPI_Abort` in deal.II `.gdbinit`?

**Comment by oneliefleft** [comment]
Just a thought. Does valgrind fail (no catch) on this too? I am away from
my machines so I can't check right now.


06.06.2017 18:14 "Wolfgang Bangerth" <notifications@github.com> napisał(a):

**Comment by Rombur** [comment]
Just for info, the `Moose` developers just tell people that they need to use `b MPI_Abort` before running their code in a debbugger.

**Comment by tjhei** [comment]
> Just a thought. Does valgrind fail (no catch) on this too? I am away from
> my machines so I can't check right now.

Same with valgrind (just exits) and it wouldn't help me anyways, because I want to be in gdb when the Assert triggers so I can inspect variables higher up in the call stack to figure out what the bug is.

> Just for info, the Moose developers just tell people that they need to use b MPI_Abort before running their code in a debbugger.

That probably works, but not ideal.

**Comment by bangerth** [comment]
For the record, I can confirm this. It only happens if the program actually uses MPI, but of course also if you run an MPI program with one process. We could lessen the impact cheaply by just calling `MPI_Abort` if we actually use more than one process. The number of people who use gdb with more than one process is probably pretty small to begin with.

But I'm not happy with this solution either because it's just too obscure a problem to document. Where would we even say that you have to add something to your `.gdbinit` file? Where would users look for this?

**Comment by bangerth** [comment]
Just out of curiosity, I use OpenMPI. Does anyone else see this with other MPI implementations? To find out, run this little program in gdb:
```
#include <deal.II/base/mpi.h>
#include <deal.II/base/exceptions.h>

using namespace dealii;

int main (int argc, char *argv[])
{
  Utilities::MPI::MPI_InitFinalize mpi_initialization(argc, argv, 1);
  Assert (false, ExcNotImplemented());
}
```

**Comment by tjhei** [comment]
> But I'm not happy with this solution either because it's just too obscure a problem to document.

We could just print a sentence to std:cerr before calling MPI_Abort. Maybe something like #4497 ?

**Comment by bangerth** [comment]
@tjhei: Can this be closed now that #4497 is merged?

**Comment by tjhei** [comment]
Yeah. I guess we don't have a better solution for now...
